### PR TITLE
Fix for db getting updated as get(Un)Processed executes

### PR DESCRIPTION
**Pull request by snarkychef**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: jobsdb/jobsdb.go [hunk 1]
```diff
@@ -1450,14 +1447,22 @@ func (jd *HandleT) constructParameterJSONQuery(table string, parameterFilters []
...
+
 /*
 * If a query returns empty result for a specific dataset, we cache that so that
 * future queries don't have to hit the DB.
 * markClearEmptyResult() when mark=True marks dataset,customVal,state as empty.
 * markClearEmptyResult() when mark=False clears a previous empty mark
  */
 
-func (jd *HandleT) markClearEmptyResult(ds dataSetT, stateFilters []string, customValFilters []string, parameterFilters []ParameterFilterT, mark bool) {
+func (jd *HandleT) markClearEmptyResult(ds dataSetT, stateFilters []string, customValFilters []string, parameterFilters []ParameterFilterT, value cacheValue, checkAndSet *cacheValue) {
```
Comment (soumyadebm) [inline jobsdb/jobsdb.go:1465]: Why make checkAndSet a pointer?

File: jobsdb/jobsdb.go [hunk 2]
```diff
@@ -1450,14 +1447,22 @@ func (jd *HandleT) constructParameterJSONQuery(table string, parameterFilters []
...
+
+const (
+	mayFindJobs cacheValue = iota
+	noJobs
+	willTryToSet
+)
+
 /*
 * If a query returns empty result for a specific dataset, we cache that so that
 * future queries don't have to hit the DB.
```
Comment (soumyadebm) [inline jobsdb/jobsdb.go:1460]: Update the comment. Also, add some comment on why we needed to have willTryToSet. If not anything, copy paste the slack message here

File: jobsdb/jobsdb.go [hunk 3]
```diff
@@ -1450,14 +1447,22 @@ func (jd *HandleT) constructParameterJSONQuery(table string, parameterFilters []
 	return fmt.Sprintf(`(%s.parameters @> '{%s}' %s)`, table, strings.Join(allKeyValues, ","), opQuery)
 }
 
+type cacheValue int
+
+const (
+	mayFindJobs cacheValue = iota
```
Comment (soumyadebm) [inline jobsdb/jobsdb.go:1453]: Nitpicking - mayFindJobs name sounds a bit weird :) Call it hasJobs?

Comment (snarkychef) [reply jobsdb/jobsdb.go:1453]: Sure.

File: jobsdb/jobsdb.go (hunk 1 above)
Comment (snarkychef) [reply jobsdb/jobsdb.go:1465]: Right now, cacheValue is of type int and I wanted to have nil support to disable CheckAndSet behaviour in the function.

File: jobsdb/jobsdb.go (hunk 2 above)
Comment (snarkychef) [reply jobsdb/jobsdb.go:1460]: Sure.

**Review by chandumlg — APPROVED** [review]

**Comment by snarkychef** [comment]
Reproduced the race issue in old code using sleep. And verified that these changes doesn't cause the issue in same scenario.

**Review by SumanthPuram — APPROVED** [review]

**Pull request merged**